src/OpenFeature/IHook.cs

```diff
@@ -12,9 +13,9 @@ namespace OpenFeature.Model
     /// </summary>
     public interface IHook
     {
-        EvaluationContext Before<T>(HookContext<T> context, IReadOnlyDictionary<string, object> hints = null);
-        void After<T>(HookContext<T> context, FlagEvaluationDetails<T> details, IReadOnlyDictionary<string, object> hints = null);
-        void Error<T>(HookContext<T> context, Exception error, IReadOnlyDictionary<string, object> hints = null);
-        void Finally<T>(HookContext<T> context, IReadOnlyDictionary<string, object> hints = null);
+        Task<EvaluationContext> Before<T>(HookContext<T> context, IReadOnlyDictionary<string, object> hints = null);
+        Task After<T>(HookContext<T> context, FlagEvaluationDetails<T> details, IReadOnlyDictionary<string, object> hints = null);
+        Task Error<T>(HookContext<T> context, Exception error, IReadOnlyDictionary<string, object> hints = null);
+        Task Finally<T>(HookContext<T> context, IReadOnlyDictionary<string, object> hints = null);

```

Comment:
I could be missing something here, but I think this makes it required to have a hook implement every stage. Hooks don't need to feature every stage. Many would only implement one: https://github.com/open-feature/spec/blob/main/specification/hooks.md#requirement-431

I think the easiest way for you to implement this is to provide a [default implementation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/default-interface-methods) which is basically a no-op. Then implementations would only implement the stage(s) they need.

This is what @justinabrahms did with the Java SDK (with TS/JS, method impls can be optional, so it's easy from the TS/JS side): https://github.com/open-feature/java-sdk/blob/main/src/main/java/dev/openfeature/javasdk/Hook.java

Reply:
Yea this was a total oversight, I think it would probably make sense to just split each hook lifecycle out into it's own interface

```csharp
public interface IBeforeHook
public interface IAfterHook
public interface IErrorHool
public interface IFinallyHook
```

Then the consumer can just implement what they want.

```csharp
public class HookDemo : IBeforeHook, IAfterHook
```

Reply:
Yea so after waking up and thinking about this more, C# added a [default implementation](https://docs.microsoft.com/en-au/dotnet/csharp/whats-new/tutorials/default-interface-methods-versions) of an interface that would allow this to work. Since we are .net standard 2.0 we don't have access to this functionality.

The best approach is to make the IHook interface an internal interface and expose an abstract class `Hook` that has a default implementation unless overriden

Reply:
Makes sense to me. Too bad default methods aren't available! This seems like a good alternative though.